velox/functions/sparksql/aggregates/DecimalSumAggregate.h

```diff
@@ -0,0 +1,449 @@
+          if (rows.hasSelections()) {
+            clearNull(group);
+          }
+          rows.applyToSelected([&](vector_size_t i) {
+            updateNonNullValue(group, sum, isEmpty);
+          });
+        }
+      }
+    } else if (decodedPartial_.mayHaveNulls()) {
+      rows.applyToSelected([&](vector_size_t i) {
+        if (decodedPartial_.isNullAt(i)) {
+          return;
+        }
+        auto decodedIndex = decodedPartial_.index(i);
+        if (!isEmptyVector->valueAt(decodedIndex) &&
+            sumVector->isNullAt(decodedIndex)) {
+          setNull(group);
+          return;
+        } else {
+          clearNull(group);
+          auto sum = sumVector->valueAt(decodedIndex);
+          auto isEmpty = isEmptyVector->valueAt(decodedIndex);
+          updateNonNullValue(group, sum, isEmpty);
+        }
+      });
+    } else {
+      if (rows.hasSelections()) {
+        clearNull(group);
+      }
+      rows.applyToSelected([&](vector_size_t i) {
+        auto decodedIndex = decodedPartial_.index(i);
+        if (!isEmptyVector->valueAt(decodedIndex) &&
+            sumVector->isNullAt(decodedIndex)) {
+          setNull(group);
+          return;
+        } else {
+          auto sum = sumVector->valueAt(decodedIndex);
+          auto isEmpty = isEmptyVector->valueAt(decodedIndex);
+          updateNonNullValue(group, sum, isEmpty);
+        }
+      });
+    }
+  }
+
+ private:
+  template <bool tableHasNulls = true>
+  inline void updateNonNullValue(char* group, TResultType value, bool isEmpty) {
+    if constexpr (tableHasNulls) {
+      exec::Aggregate::clearNull(group);
+    }
+    auto decimalSum = accumulator(group);
+    decimalSum->overflow +=
+        DecimalUtil::addWithOverflow(decimalSum->sum, value, decimalSum->sum);
+    decimalSum->isEmpty &= isEmpty;
+  }
+
+  template <bool tableHasNulls = true>
+  inline void mergeAccumulators(char* group, DecimalSum other) {
+    if constexpr (tableHasNulls) {
+      exec::Aggregate::clearNull(group);
+    }
+    auto decimalSum = accumulator(group);
+    decimalSum->mergeWith(other);
+  }
+
+  inline DecimalSum* accumulator(char* group) {
+    return exec::Aggregate::value<DecimalSum>(group);
+  }
+
+  DecodedVector decodedRaw_;
+  DecodedVector decodedPartial_;
+  TypePtr sumType_;
+};
+
+exec::AggregateRegistrationResult registerDecimalSumAggregate(
+    const std::string& name) {
+  std::vector<std::shared_ptr<exec::AggregateFunctionSignature>> signatures{
+      exec::AggregateFunctionSignatureBuilder()
+          .integerVariable("a_precision")
+          .integerVariable("a_scale")
+          .integerVariable("r_precision", "min(38, a_precision + 10)")
+          .integerVariable("r_scale", "min(38, a_scale)")
+          .argumentType("DECIMAL(a_precision, a_scale)")
+          .intermediateType("ROW(DECIMAL(r_precision, r_scale), boolean)")
+          .returnType("DECIMAL(r_precision, r_scale)")
+          .build()};
```

Comment:
We may need to register all supported signatures for spark sum function. Otherwise, other types' support is missing. Here are my changes [registerSumAggregate](https://github.com/oap-project/velox/commit/51a05a1844d670f754a8e98a14ebaafb6f809f7a#diff-454ca54c7824ad31f30826bf27e5e3053f62b77f668fbb3f1ea7a1c0f392ed9dR382-R509).

Reply:
I see, the original method was to register a new decimal_sum function, but now we need to use the same sum function.

Reply:
@rui-mo I create a new PR for this change, https://github.com/facebookincubator/velox/pull/6889.

Reply:
It seems that the signatures of an aggregate function need to be registered at one time, and we cannot partially overwrite them. With above change, I suspect that other types' registration will still be missing if we only register sum for decimal type in spark sql.
Link for reference: [Aggregate.cpp](https://github.com/facebookincubator/velox/blob/main/velox/exec/Aggregate.cpp#L65-L77).

Reply:
Makes sense, thanks.